sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,549 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/testing/usb_logging.h"
+
+#include <ctype.h>
+
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/print.h"
+#include "sw/device/lib/testing/pinmux_testutils.h"
+#include "sw/device/lib/testing/usb_testutils.h"
+#include "sw/device/lib/testing/usb_testutils_controlep.h"
+
+#define MODULE_ID MAKE_MODULE_ID('u', 's', 'l')
+
+// Size of each logging buffer; data is collected in a buffer until the buffer
+// is full (or a timeout occurs) and is then transmitted as a single transfer
+// at higher throughput.
+enum {
+  kUSBLogBufferSize = 0x800u,
+};
+
+// Total number of logging buffers to be shared among all active streams.
+enum {
+  kUSBLogNumBuffers = 8u,
+};
+
+// Maximum delay before logging data is flushed upstream to the host;
+// microseconds.
+enum {
+  kUSBLogFlushInterval = 500u * 1000u,
+};
+
+// Maximum length of the configuration descriptor.
+enum {
+  kCfgDscrLenMax =
+      (USB_CFG_DSCR_LEN +
+       kUSBLogMaxStreams * (USB_INTERFACE_DSCR_LEN + 1u * USB_EP_DSCR_LEN)),
+};
+
+// Forward reference required for linked list.
+typedef struct usb_logging_buf usb_logging_buf_t;
+
+// USB logging buffer description.
+struct usb_logging_buf {
+  /**
+   * Link to the next buffer for this stream, or the next free buffer.
+   */
+  usb_logging_buf_t *next;
+  /**
+   * Number of used bytes; index at which next write shall occur.
+   */
+  uint16_t bytes_used;
+  /**
+   * Data buffer holding logging data.
+   */
+  uint8_t data[kUSBLogBufferSize];
+};
+
+// Forwards reference to logging context.
+typedef struct usb_logging_ctx usb_logging_ctx_t;
+
+// Per-stream state information.
+typedef struct usb_logging_stream {
+  /**
+   * Context pointer; for use within callback handler.
+   */
+  usb_logging_ctx_t *ctx;
+  /**
+   * Reliable delivery required?
+   */
+  bool reliable;
+  /**
+   * Remapping for non-printable characters required?
+   */
+  bool remap;
+  /**
+   * The USB device endpoint that we're using.
+   */
+  uint8_t ep;
+  /**
+   * Currently sending a buffer?
+   */
+  bool sending;
```

Comment:
You probably want to pack this struct better.

Reply:
Nevermind this. It seems when I wrote this comment I didn't properly context switch and was assuming a 64-bit target ;)

Reply:
Offline: 8-bit quantities are already grouped.